crates/worker/src/partition/leadership.rs

```diff
@@ -673,76 +727,294 @@ where
+                ),
+            ))),
+            State::Leader(leader_state) => {
+                match leader_state
+                    .self_proposer
+                    .propose_with_notification(partition_key, cmd)
+                    .await
+                {
+                    Ok(commit_token) => {
+                        leader_state
+                            .awaiting_rpc_self_propose
+                            .push(SelfAppendFuture(commit_token, Some(reciprocal)));
+                    }
+                    Err(e) => {
+                        respond_to_rpc(
+                            reciprocal
+                                .prepare(Err(PartitionProcessorRpcError::Internal(e.to_string()))),
                         );
                     }
-                    Ok(())
                 }
-            },
-        );
+            }
+        }
+    }
+}
+
+fn respond_to_rpc(
+    outgoing: Outgoing<
+        Result<PartitionProcessorRpcResponse, PartitionProcessorRpcError>,
+        HasConnection,
+    >,
+) {
+    // ignore shutdown errors
+    let _ = task_center().spawn(
+        TaskKind::Disposable,
+        "partition-processor-rpc-response",
+        None,
+        async move { outgoing.send().await.map_err(Into::into) },
+    );
+}
 
-        if maybe_task.is_err() {
-            let invocation_id_str = invocation_id
-                .as_ref()
-                .map(|i| i.to_string())
-                .unwrap_or_default();
-            trace!(
-                restate.invocation.id = %invocation_id_str,
-                "Partition processor is shutting down, we are not sending the message to ingress",
+struct SelfAppendFuture(
+    CommitToken,
+    Option<Reciprocal<Result<PartitionProcessorRpcResponse, PartitionProcessorRpcError>>>,
+);
+
+impl SelfAppendFuture {
+    fn fail_with_internal(&mut self) {
+        if let Some(reciprocal) = self.1.take() {
+            respond_to_rpc(reciprocal.prepare(Err(PartitionProcessorRpcError::Internal(
+                "error when proposing to bifrost".to_string(),
+            ))));
+        }
+    }
+
+    fn fail_with_not_leader(&mut self, this_partition_id: PartitionId) {
+        if let Some(reciprocal) = self.1.take() {
+            respond_to_rpc(
+                reciprocal.prepare(Err(PartitionProcessorRpcError::NotLeader(
+                    this_partition_id,
+                ))),
             );
         }
+    }
+}
+
+impl Future for SelfAppendFuture {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let append_result = ready!(self.0.poll_unpin(cx));
+
+        if append_result.is_err() {
+            self.get_mut().fail_with_internal();
+            return Poll::Ready(());
+        }
+        respond_to_rpc(
+            self.get_mut()
+                .1
+                .take()
+                .expect("Future should be called once")
```

Comment:
Could be a `if let` instead of a `take` to cover the case that we already sent an error.